src/actions/gifs.js

```diff
@@ -0,0 +1,604 @@
+                    }
+                } else {
+                    dispatch(updateSearchText(tagName));
+                    dispatch(cacheGifsRequest(json.gfycats));
+                    dispatch(receiveCategorySearch({tagName, json}));
+
+                    dispatch(sendViewCountBatch({
+                        gfycats: json.gfycats,
+                        params: {
+                            context: 'category',
+                            keyword: tagName,
+                            app_id: getState().entities.gifs.app.appId,
+                        },
+                    }));
+
+                    // preload categories list
+                    if (tagName === 'trending') {
+                        dispatch(requestCategoriesListIfNeeded());
+                    }
+                }
+            }
+        ).catch((err) => dispatch(errorSearching(err, tagName)));
+    };
+}
+
+export function shouldSearch(state, searchText) {
+    const resultsByTerm = state.search.resultsByTerm[searchText];
+    if (!resultsByTerm) {
+        return true;
+    } else if (resultsByTerm.isFetching) {
+        return false;
+    } else if (resultsByTerm.moreRemaining) {
+        return true;
+    }
+    return resultsByTerm.didInvalidate;
+}
+
+export function searchIfNeeded(searchText) {
+    return (dispatch, getState) => {
+        if (shouldSearch(getState(), searchText)) {
+            if (searchText.toLowerCase() === 'trending') {
+                return dispatch(searchCategory({tagName: searchText}));
+            }
+            return dispatch(searchGfycat({searchText}));
+        }
+        return Promise.resolve();
+    };
+}
+
+export function searchIfNeededInitial(searchText) {
+    return (dispatch, getState) => {
+        dispatch(updateSearchText(searchText));
+        if (shouldSearchInitial(getState(), searchText)) {
+            if (searchText.toLowerCase() === 'trending') {
+                return dispatch(searchCategory({tagName: searchText}));
+            }
+            return dispatch(searchGfycat({searchText}));
+        }
+        return Promise.resolve();
+    };
+}
+
+export function shouldSearchInitial(state, searchText) {
+    const resultsByTerm = state.entities.gifs.search.resultsByTerm[searchText];
+    if (!resultsByTerm) {
+        return true;
+    } else if (resultsByTerm.isFetching) {
+        return false;
+    }
+    return false;
+}
+
+export function searchById(gfyId) {
+    return (dispatch) => {
+        dispatch(requestSearchById(gfyId));
+        return gfycatSdk.searchById({id: gfyId}).then(
+            (response) => {
+                dispatch(receiveSearchById(gfyId, response.gfyItem));
+                dispatch(cacheGifsRequest([response.gfyItem]));
+            }
+        ).catch((err) => dispatch(errorSearchById(err, gfyId)));
+    };
+}
+
+export function shouldSearchById(state, gfyId) {
+    const gif = state.cache.gifs[gfyId];
+    if (!gif) {
```

Comment:
This could be simplified to `return !gif;`

Reply:
I've changed to `return !state.cache.gifs[gfyId]`